scripts/metamath-extract.py

```diff
@@ -36,86 +53,151 @@ def compressed_proof_get_lemmas(proof: str) -> tuple[str, ...]:
     return tuple(proof[lemmas_begin:lemmas_end].split())
 
 
-def theorem_get_disjoints(theorem: Theorem) -> tuple[DisjointStatement, ...]:
-    ret = []
-    for disjoint in theorem.context.disjoints:
-        new_vars = tuple(
-            Metavariable(var) for var in theorem.statement.get_metavariables() if var in disjoint.get_metavariables()
-        )
-        if new_vars:
-            ret += [DisjointStatement(new_vars)]
-    return tuple(ret)
-
-
-def copy_with_context(
-    hypotheses: tuple[Statement, ...], conclusion: StructuredStatement, output_composer: Composer
-) -> None:
-    with output_composer.new_context():
-        for hypothesis in hypotheses:
-            output_composer.load(hypothesis)
-        output_composer.load(conclusion)
-
-
-def extract_proof(input_composer: Composer, output_composer: Composer, theorem_name: str) -> None:
-    theorem = input_composer.get_theorem(theorem_name)
-    assert isinstance(theorem.statement, ProvableStatement)
-    proof = theorem.statement.proof
-    assert proof, 'Theorem has no proof.'
-
-    lemmas: tuple[Theorem, ...] = tuple(
-        input_composer.get_theorem(lemma) for lemma in compressed_proof_get_lemmas(proof)
-    )
-
-    mm_symbols: set[str] = {'(', ')'}
-    variables: set[str] = set()
-    for t in (*lemmas, theorem):
-        for statement in (
-            *t.context.essentials,
-            *t.context.floatings,
-            t.statement,
+def supporting_database_for_provable(
+    cut_antecedents: dict[str, FloatingStatement | AxiomaticStatement | Block],
+    global_disjoints: set[frozenset[str]],
+    provable: ProvableStatement,
+    essentials: tuple[DisjointStatement | EssentialStatement, ...],
+) -> Database:
+    statements: list[Statement] = []
+
+    proof = provable.proof
+    assert proof, f'Proof is missing for {provable.label}'
+    needed_lemmas = compressed_proof_get_lemmas(proof)
+
+    needed_constants = set()
+    needed_metavariables = set()
+    for needed in (provable, *essentials, *(cut_antecedents[lemma_name] for lemma_name in needed_lemmas)):
+        needed_constants.update(statements_get_constants((needed,)))
+        needed_metavariables.update(needed.get_metavariables())
+    statements.append(ConstantStatement(tuple(needed_constants)))
+    if needed_metavariables:
+        statements.append(VariableStatement(tuple(Metavariable(var) for var in needed_metavariables)))
+
+    for pair in global_disjoints:
+        if pair.issubset(needed_metavariables):
+            statements.append(DisjointStatement(tuple(Metavariable(var) for var in pair)))
+
+    for lemma_name, lemma_statement in cut_antecedents.items():
```

Comment:
Wasn't the problem with Floating statements they are order sensitive? Is iterating over `cut_antecedents` in the same order as the order of floating statements in the original db? This seems fishy, as it is a `dict`.

Reply:
Yeah, since python 3.7, iterating over dicts is in order of insertion.

https://docs.python.org/3.7/library/stdtypes.html#dictionary-view-objects